File: api/create-bill.js
// 这是一个在Vercel上运行的Node.js无服务器函数

// TODO: 这是一个示例函数，在真实应用中，您需要从您的用户认证系统
// (例如通过解析cookie或token)来获取当前登录的用户信息。
// 目前，我们先用一个固定的虚拟用户数据来完成流程。
const getCurrentUser = async () => {
  return { id: 'usr_12345', email: 'customer@example.com', name: 'John Doe' };
};

// Vercel 会将这个文件作为一个API端点来处理
export default async function handler(req, res) {
  // 首先，检查请求方法是否为POST，确保安全
  if (req.method !== 'POST') {
    return res.status(405).json({ message: 'Only POST requests are allowed' });
  }

  try {
    // 从Vercel环境变量中安全地读取我们的“钥匙”
    const apiKey = process.env.BILLPLZ_API_KEY;
    const collectionId = process.env.BILLPLZ_COLLECTION_ID;
    const appUrl = process.env.APP_URL;

    // 从前端发来的请求体(body)中，获取用户选择的套餐信息
    const { planId, planName, amount } = req.body;

    // 获取当前用户信息
    const user = await getCurrentUser();
    if (!user) {
      return res.status(401).json({ message: 'Authentication required' });
    }
    
    // 准备要发送给Billplz的数据包 (Payload)
    const billplzPayload = {
      collection_id: collectionId,
      email: user.email,
      name: user.name,
      amount: amount, // 从前端接收到的、以仙为单位的价格
      description: `PF Creative - ${planName}`,
      // 关键部分：设置回调和跳转URL
      callback_url: `${appUrl}/api/webhook-billplz`,
      redirect_url: `${appUrl}/payment-success`,
      // 将我们的内部用户ID传递给Billplz，以便后续识别
      reference_1_label: 'UserID',
      reference_1: user.id 
    };

    // 使用fetch函数，向Billplz的Staging(测试)环境API发送请求
    const billplzResponse = await fetch('https://www.billplz.com/api/v3/bills', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        // Billplz需要使用HTTP Basic Auth进行认证
        'Authorization': `Basic ${Buffer.from(apiKey + ':').toString('base64')}`
      },
      body: JSON.stringify(billplzPayload)
    });

    // 检查Billplz的响应是否成功
    if (!billplzResponse.ok) {
      const errorData = await billplzResponse.json();
      // 在服务器端打印详细错误，方便调试
      console.error('Billplz API Error:', errorData);
      throw new Error('Failed to create a bill with Billplz.');
    }

    const billplzResult = await billplzResponse.json();

    // 如果一切顺利，将从Billplz获取到的支付链接URL，以JSON格式返回给前端
    res.status(200).json({ url: billplzResult.url });

  } catch (error) {
    // 如果过程中出现任何错误，在服务器端打印错误，并向前端返回一个通用错误信息
    console.error('Create Bill API Error:', error);
    res.status(500).json({ message: error.message || 'An internal server error occurred.' });
  }
}

File: api/create-bill.py
import json
import jwt
import base64
import urllib.request
import urllib.error
import logging
import os
import psycopg2
import psycopg2.pool

# --- START: Added self-contained functions and constants ---
# This logic is copied from your utils.py to make this file independent.

logging.basicConfig(level=logging.INFO)

db_pool = None

# Vercel Environment variables will be automatically loaded
BILLPLZ_API_KEY = os.environ.get('BILLPLZ_API_KEY', "f9478c0c-a6fc-444b-9132-69b144a7af47")
BILLPLZ_COLLECTION_ID = os.environ.get('BILLPLZ_COLLECTION_ID', "ek0rvdud")
BASE_URL = os.environ.get('BASE_URL', "https://pfcreativestudio.vercel.app")
JWT_SECRET = os.environ.get('JWT_SECRET', 'd528502f7a76853766814ffd7bdad0d3577ef4c1273995402a2239493a5d19cd')

def init_connection_pool():
    global db_pool
    if db_pool:
        return
    try:
        db_user = os.environ.get('DB_USER')
        db_password = os.environ.get('DB_PASSWORD')
        db_name = os.environ.get('DB_NAME')
        instance_connection_name = os.environ.get('INSTANCE_CONNECTION_NAME')
        db_socket_dir = '/cloudsql'
        conn_string = f'user={db_user} password={db_password} dbname={db_name} host={db_socket_dir}/{instance_connection_name}'
        db_pool = psycopg2.pool.SimpleConnectionPool(1, 5, dsn=conn_string)
        logging.info("Database connection pool initialized successfully for create-bill.")
    except Exception as e:
        logging.error(f"Error initializing connection pool for create-bill: {e}")
        db_pool = None

init_connection_pool()

def get_db_connection():
    if not db_pool:
        raise Exception("Database connection pool is not available.")
    return db_pool.getconn()

def release_db_connection(conn):
    if db_pool:
        db_pool.putconn(conn)

def cors_preflight(request):
    if request.method == 'OPTIONS':
        headers = {
            'Access-Control-Allow-Origin': '*',
            'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
            'Access-Control-Allow-Headers': 'Content-Type, Authorization, X-Admin-Password',
            'Access-Control-Max-Age': '3600'
        }
        return ('', 204, headers)
    return None

def get_cors_headers():
    return {
        'Access-Control-Allow-Origin': '*',
        'Content-Type': 'application/json'
    }

# --- END: Added self-contained functions and constants ---


def handler(request):
    """
    Handles the creation of a payment bill with Billplz.
    """
    preflight_response = cors_preflight(request)
    if preflight_response:
        return preflight_response

    headers = get_cors_headers()
    conn = None
    cursor = None # Define cursor here
    try:
        auth_header = request.headers.get('Authorization')
        if not auth_header or not auth_header.startswith('Bearer '):
            return (json.dumps({'error': 'Authorization token required'}), 401, headers)
        
        token = auth_header.split(' ')[1]
        payload = jwt.decode(token, JWT_SECRET, algorithms=['HS256'])
        username = payload.get('username')
        if not username:
            return (json.dumps({'error': 'Invalid token: username missing'}), 401, headers)

        conn = get_db_connection()
        cursor = conn.cursor()
        cursor.execute('SELECT username FROM users WHERE username = %s', (username,))
        user_record = cursor.fetchone()
        if not user_record:
            return (json.dumps({'error': 'User not found'}), 404, headers)

        request_json = request.get_json(silent=True) or {}
        plan_name = request_json.get('planName')
        amount = request_json.get('amount')
        plan_id = request_json.get('planId', '')
        
        if not plan_name or not amount:
            return (json.dumps({'error': 'Plan information is missing'}), 400, headers)

        billplz_payload = {
            'collection_id': BILLPLZ_COLLECTION_ID,
            'email': f"{username}@pfcreative.system",
            'name': username,
            'amount': str(int(amount)),
            'description': plan_name[:200],
            'callback_url': f"{BASE_URL}/api/webhook-billplz",
            'redirect_url': f"{BASE_URL}/payment-success.html",
            'reference_1_label': 'Username',
            'reference_1': username,
            'reference_2_label': 'PlanID',
            'reference_2': plan_id
        }
        
        req_headers = {
            'Content-Type': 'application/json',
            'Authorization': f'Basic {base64.b64encode(f"{BILLPLZ_API_KEY}:".encode()).decode()}'
        }
        
        req = urllib.request.Request(
            'https://www.billplz.com/api/v3/bills',
            method='POST',
            data=json.dumps(billplz_payload).encode('utf-8'),
            headers=req_headers
        )

        with urllib.request.urlopen(req) as response:
            if response.getcode() == 200:
                billplz_result = json.loads(response.read().decode())
                if 'url' in billplz_result:
                    logging.info(f"Payment link created for {username}: {billplz_result['url']}")
                    return (json.dumps({
                        'url': billplz_result['url'],
                        'bill_id': billplz_result.get('id', '')
                    }), 200, headers)
            
            error_body = response.read().decode()
            logging.error(f"Billplz API error: {response.getcode()} - {error_body}")
            return (json.dumps({'error': 'Failed to create payment link from provider'}), 502, headers)

    except urllib.error.HTTPError as e:
        error_body = e.read().decode()
        logging.error(f"Billplz HTTP Error: {e.code} - {error_body}")
        return (json.dumps({'error': 'Payment service provider error'}), 502, headers)
    except urllib.error.URLError as e:
        logging.error(f"Billplz connection error: {str(e)}")
        return (json.dumps({'error': 'Cannot connect to payment service'}), 503, headers)
    except Exception as e:
        logging.error(f"Create Bill Error: {str(e)}", exc_info=True)
        return (json.dumps({'error': 'Internal server error'}), 500, headers)
    finally:
        if conn:
            if cursor:
                try:
                    cursor.close()
                except Exception as e:
                    logging.error(f"Error closing cursor: {e}")
            release_db_connection(conn)
